File: python/trapping_rain_water.py
#no. 42
class Solution:
    def trap(self, height: List[int]) -> int:
        stack = [0]
        ans = 0
        for i in range(1, len(height)):
            while stack and (height[i] > height[stack[-1]]):
                if len(stack) >= 2:
                    ans += (min(height[stack[-2]], height[i]) - height[stack[-1]]) * (i - stack[-2] - 1)
                stack.pop()
            stack.append(i)
        return ans

File: C++/trapping_rain_water.cpp
// no. 42
#include <vector>
using namespace std;
class Solution {
public:
  int trap(vector<int> &height) {
    vector<int> stack;
    stack.push_back(0);
    int ans = 0;
    for (int i = 0; i < height.size(); i++) {
      while (height[i] > height[stack.back()]) {
        if (stack.size() >= 2) {
          int h = min(height[stack[stack.size() - 2]], height[i]) -
                  height[stack.back()];
          int w = (i - stack[stack.size() - 2] - 1);
          ans += h * w;
        }
        stack.pop_back();
        if (stack.empty()) {
          break;
        }
      }
      stack.push_back(i);
    }
    return ans;
  }
};
